keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,204 @@
+
+ ## Proposal
+
+ ### User Stories
+
+ As an application developer, I want my application pods to be scheduled onto
+one node that has pod with the "foo" tag and the tag value between "20" and "30".
+
+- if we use the original `In` operator to implement, then we will write affinity 
+like this:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ This is not an ideal solution. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to understand the impact of these changes, we need to understand the their performance implications. We will perform performance/benchmark tests on this change and backward compatibility test.
+
+ Compared to the time that inter-pod affinity was introduced, performance of the scheduler has been greatly improved(https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
+ ## Design Details
+ ### Content
+  We will abstract the `Node` and `Pod` label selection feature, and finally the `NodeSelector` and `PodSelector` will be based on the `NodeAndPodSelectorRequirement` implementation:
+
+ The definition of `NodeSelector` will not change,it is defined as below:
+ ```go
+type NodeSelector struct {
+  NodeSelectorTerms []NodeSelectorTerm
+}
+```
+
+For the definition of `NodeSelectorTerm`, today the api looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []NodeSelectorRequirement
+  MatchFields []NodeSelectorRequirement
+}
+```
+
+We will use `NodeAndPodSelectorRequirement` to replace the original `NodeSelectorRequirement` as the base selector of `NodeSelector`. So it looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []NodeAndPodSelectorRequirement
```

Comment:
Name it by what it does, not where it's used. SchedulerSupportedSelectorRequirement? NumericAwareSelectorRequirement? I have to run, maybe @liggitt can think of a better name.

NodeAndPod won't work for this, the name can't be about the things being selected, it needs to be about the selection process. Otherwise people will probably use the type wrongly.

Reply:
Sorry, please forgive my doubts. I don't particularly understand why mentioned numeric. `Gt` or` Lt` is not only able to compare numbers but also to compare letters, in other words it compares strings (including numbers).

I think `SchedulerSupportedSelectorRequirement` may be a possible choice, but inspired by you, I think `SchedulerSelectorRequirement` is more concise and appropriate. What do you think of it?

Reply:
> `Gt` or` Lt` is not only able to compare numbers but also to compare letters, in other words it compares strings (including numbers).

that is inconsistent with how these are used today:

https://github.com/kubernetes/kubernetes/blob/8ec6167f61463cbc6874514b02d0137d51b23542/staging/src/k8s.io/apimachinery/pkg/labels/selector.go#L151-L159

Are there compelling use cases for supporting lexical comparisons? It certainly makes the behavior less intuitive... e.g. what would the sort order of the following be?
```
1
1a
2
2a
10
20
```

Reply:
This conversation is exactly why it needs to be prefixed with either "Numeric" or "Lexical" (or if it's some hybrid, something explaining that). We shouldn't make users work hard to figure this out.

Lexical comparison implies that "10" < "2", I do not think users will want that behavior.

Numeric comparison implies that error handling is needed if either of the label values are not numbers.

Reply:
+1 to numeric comparison + error checking. I don't think there are many users who need lexical comparison in the API.

Reply:
Ok. I mean, in fact, we use `NumericXXX` is not suitable, because there are still operators such as "In", "Equals" and so on.

Therefore, I agree that Gt and Lt only work on numbers. Summarize the above discussion:


1. `Gt` and `Lt` only work on numbers
2. `NodeAndPodSelectorRequirement` changed to `SchedulerSelectorRequirement`

What do you think of this?
@lavalamp @liggitt @bsalamat

Reply:
In, Equals, etc all would behave the same whether the type is string or numeric. It's only ordered comparisons where the distinction becomes important.

"SchedulerSelectorRequirement" is not my favorite name, it's named for the component you expect to consume the thing. It is better than "NodeAndPod", but still: other components might eventually interpret this, the name isn't future-proof. It's also not at all obvious that "scheduler" here means "supports ordered comparisons on numeric values". That's really what you want the name to imply to people.

Reply:
Understood. 

Naming is really the hardest thing, I can't think of a better name than `SchedulerSupportedNumericSelectorRequirement`. Do you have any advice?

Reply:
Some ideas:

* NumericAwareSelectorRequirement
* NumericallyOrderedSelectorRequirement
* AffinitySelectorRequirement (only if this is going to be used for all and only affinity selection features)

Reply:
Agree to use NumericAwareSelectorRequirement, I have updated. Thanks a lot for comments and your review.